extensions/ql-vscode/src/queries-panel/query-pack-discovery.ts

```diff
@@ -0,0 +1,93 @@
+import { basename, dirname } from "path";
+import { CodeQLCliServer, QuerySetup } from "../codeql-cli/cli";
+import { Event } from "vscode";
+import { App } from "../common/app";
+import { getOnDiskWorkspaceFolders } from "../helpers";
+import { dbSchemeToLanguage } from "../common/query-language";
+import { FALLBACK_QLPACK_FILENAME, QLPACK_FILENAMES } from "../pure/ql";
+import { FilePathDiscovery } from "./file-path-discovery";
+import { getErrorMessage, isEqualOrHigherPath } from "../pure/helpers-pure";
+import { extLogger } from "../common";
+import { EOL } from "os";
+
+export interface QueryPack {
+  path: string;
+  language: string | undefined;
+}
+
+/**
+ * Discovers all query packs contained in the workspace.
+ */
+export class QueryPackDiscovery extends FilePathDiscovery<QueryPack> {
+  constructor(app: App, private readonly cliServer: CodeQLCliServer) {
+    super("Query Pack Discovery", "**/{qlpack.yml,codeql-pack.yml}", app);
```

Comment:
Does this capture the scenario when a new pack file is added to the workspace?

Reply:
I believe so. The watcher is installed on every workspace folder and not just those that currently contain query packs. It then triggers whenever a file changes that matches this pattern, as well as whenever a directory changes.

I'll make sure to test this in all the ways I can think of.